### ISSUE TITLE
wintc-taskband support for other themes (Royale + Royale Noir, Zune, etc)

### ISSUE THREAD

**Issue by enderman0125** [rozniak/xfce-winxp-tc#352]
Any possibility of updating wintc-taskband to support other themes in the pack such as Royale Noir and Zune? It looks a bit out of place when I have Royale Noir applied but the taskbar is the regular blue Luna theme. Is this planned in the future?

**Comment by rozniak** [comment]
See the `additional-theme` label: https://github.com/rozniak/xfce-winxp-tc/labels/additional-theme

At the moment only the XFWM4 decorations are present for those other themes, the GTK stuff isn't done yet.

**Comment by enderman0125** [comment]
Ah, alright.

**Comment by acdavit** [comment]
I'm wondering if it's possible to make an automated script that would take preexisting XP themes and convert them to something that both XFCE4 and winxp-tc could use, There are like thousands of unofficial Windows XP themes floating around on the internet that I would love to use on my XP-fied Linux installation

To what extent are the bitmaps from the original Windows XP themes altered? Would it even be possible to do something similar to this or would there be too much voodoo involved?

**Comment by rozniak** [comment]
@acdavit I have thought about that for quite some time, I think the last time I commented on the idea was in #222 . I did recently start toying around with it in a branch `ft-uxtheme`, essentially all that's there at the moment is the beginnings of a Python tool for extracting the theme resources as a test. Trouble I came across instantly was that PIL/Pillow does not correctly support 32bpp bitmaps (used by some theme parts like the radio button). I didn't really feel like faffing with parsing the BITMAPV3 headers or whatever manually right now so it's left where it is - if you know some Python and feel like chipping in there it would be useful.

For the time being though I figured it would be best to get the themes done the same way as I have done Luna (Blue) and Professional, just so that there's at least something rather than nothing.

I should probably write my thoughts on what might be needed in a separate issue. Just in general a good start would be getting the resources extracted correctly without losing the alpha channel from the 32bpp bitmaps.

**Comment by acdavit** [comment]
@rozniak 

Do you mind if I post a question about it on stack overflow?

**Comment by rozniak** [comment]
@acdavit About parsing those 32bpp bitmaps? It's a bit odd, I had a poke in the PIL/Pillow source and it suggests that these headers are supported: https://github.com/python-pillow/Pillow/blob/main/src/PIL/BmpImagePlugin.py#L102

Not really sure why the alpha channel is lost though.

**Comment by acdavit** [comment]
@rozniak 

Exactly. Would you be okay if I posted the code on stack overflow or Reddit? Maybe someone could shed some light on this mystery.

**Comment by acdavit** [comment]
@rozniak 

Thanks for the response. 

I did submit a question to stack exchange, but I also noticed something bizarre while toying around with the resources. 

Resource Hacker on Windows does seem to show thumbnails of the alpha channel correctly, but the extracted bitmap lacks it, just like how it does on our Python script. 

[Screen-Recording(1).webm](https://github.com/user-attachments/assets/1bec45bc-c5e1-4ecd-807d-dad4b0b24a4f)

Could this, yet again, be the fault of Resource Hacker incorrectly parsing 32bpp bitmaps, or could there be something fishy about how transparency is handled in msstyles?

I'm curious to hear your thoughts on this matter.

**Comment by rozniak** [comment]
@acdavit Yeah that's why I don't use Resource Hacker for the theme work. :stuck_out_tongue:

**Comment by acdavit** [comment]
@rozniak 

In the fashion of Stack Overflow, my post was removed for being "a duplicate" and referred me to [this post](https://stackoverflow.com/questions/10453604/load-rgba-bitmap-with-pil). 

I did try incorporating this code into yours, but no matter what I tried, it always errors out with "Syntax Error: Not a BMP file".

**Comment by BenBE** [comment]
Have you tried reading the resource section from the binary in question directly? From there parsing out the actual Bitmap resource is kinda straight forward.

**Comment by rozniak** [comment]
I have dumped out the resources raw before and was a bit bamboozled by the fact that they don't appear to have BMP header (have a look in a hex editor, the signature for a bitmap file isn't there).

Out of curiousity I had a play with Pillow a bit more, noted that it actually is reading using the `DIB` format, and not `BMP`.

I'm going to see if it's viable to prepend a bitmap header onto the data, and then see if Pillow is happy to read it as a 32bpp bitmap.

**Comment by rozniak** [comment]
I'm fairly certain this is a bug with Pillow tbh. Basically the resources here are DIBs with the 40 byte header (`BITMAPINFOHEADER`), and `BI_RGB` as the compression (aka uncompressed). I did alter the script to reconstruct the missing BMP file header, just in case it wasn't being read properly by Pillow - I still get the same results though of no alpha channel.

**Comment by babaisgit** [comment]
what modifications would i have to make to implement this? i'd like to use this full time with the zune theme, i figured i'd just do it myself.

**Comment by rozniak** [comment]
@babaisgit For now you can ignore most of the above conversation, because it relates to trying to automate MSSTYLES --> XFWM4/GTK3 themes.

For doing it manually, you probably want to copy the Silver theme over and start by modifying that. aka copy `/themes/luna/metallic` into `/themes/zune` (but do not overwrite the `xfwm4` directory, because that's already done.

If you have a look at this commit: https://github.com/rozniak/xfce-winxp-tc/commit/2d1e0689b5224eec511e159ac39b996ea1636465

You will see I made Silver by copying and modifying the Blue theme. The metrics are 99% the same (margins, padding, etc.) in all themes, so the majority of the work is essentially just replacing the graphics in the `Resources` dir.

If you want to talk more about it, ask questions etc. then probably best to continue in #9  :wink:

**Comment by rozniak** [comment]
@acdavit @BenBE Good news - I have just had a little poke with the PIL source I linked and yeah, removing that `header == 22` check completely fixes the issue. I'm not sure there's a good reason for it, I think it's probably the usual thing where people are only ever interested in ripping icons and cursors and nothing else. I will try and PR a fix upstream.

**Comment by rozniak** [comment]
Solution I may propose to PIL is adding a couple extra 'formats', `DIBA` and `BMPA` that pretty much just 'hint' that the alpha channel should be considered. Dunno if upstream will accept it but I'll write up a PR anyway and give it a go.

**Comment by rozniak** [comment]
Okay so I raised this issue upstream with Pillow here: https://github.com/python-pillow/Pillow/issues/8594

Not sure that my workaround with the extra formats will be worth a merge there, it's up to them. The good news is though is that the developer that responded to my issue was extremely helpful and provided a Python snippet that would read in the alpha channel, working with current Pillow.

I have scrapped my old weird hacky bodgefest and merged this snippet into the `ft-uxtheme` branch. If you run `themgood.py` against a `.msstyles` file now it should dump both the raw DIB and a corresponding RGBA PNG. We are in business. :clinking_glasses:

**Comment by rozniak** [comment]
I think for the tool it needs to be a kind of two phase process:
- First phase to extract out the data from msstyles, DIB->PNG, gather up all the metrics
- Second phase to generate the themes for each part of the DE -- XFWM, GTK (2, 3, ...4?), Qt(?)

**Comment by rozniak** [comment]
Discussion about MSSTYLES should probably continue in #222 or something, the themes mentioned in OP (Royale Noir, Zune) have been implemented.

**Comment by dominichayesferen** [comment]
> [@rozniak](https://github.com/rozniak)
> 
> Thanks for the response.
> 
> I did submit a question to stack overflow, but I also noticed something bizarre while toying around with the resources.
> 
> Resource Hacker on Windows does seem to show thumbnails of the alpha channel correctly, but the extracted bitmap lacks it, just like how it does on our Python script.
> 
>  Screen-Recording.1.webm 
> Could this, yet again, be the fault of Resource Hacker incorrectly parsing 32bpp bitmaps, or could there be something fishy about how transparency is handled in msstyles?
> 
> I'm curious to hear your thoughts on this matter.

It's worth also noting that msstyles have a `Transparency` value, optionally, that is used to key out the exact colour when rendering in Win32 - e.g.: Royale tray has red as its value, matching with its red tray extension portion.

Not every case does - I presume it falls back to being pink for obvious reasons - but, yeah.
